File: Top 100 Codes/04-Important Codes related to Arrays/28-Rotation of elements of array- left and right/Rotationofelementsofarrayleftandright.py
def reverse(arr, start, end):
    while start < end:
        arr[start], arr[end] = arr[end], arr[start]
        start += 1
        end -= 1

def rotate_left(arr, d):
    n = len(arr)
    d %= n  # Handle cases where d > n
    reverse(arr, 0, d - 1)
    reverse(arr, d, n - 1)
    reverse(arr, 0, n - 1)

def rotate_right(arr, d):
    rotate_left(arr, len(arr) - (d % len(arr)))  # Right rotation is equivalent to left rotation by (n-d)

# Example usage
arr = [1, 2, 3, 4, 5, 6, 7]
d = int(input("Enter number of positions to rotate: "))

left_arr = arr[:]  # Copy original array for left rotation
right_arr = arr[:]  # Copy original array for right rotation

rotate_left(left_arr, d)
rotate_right(right_arr, d)

print("Left Rotation:", left_arr)
print("Right Rotation:", right_arr)

File: Top 100 Codes/04-Important Codes related to Arrays/28-Rotation of elements of array- left and right/Rotationofelementsofarrayleftandright.cpp
#include <iostream>
using namespace std;

// Function to reverse a portion of the array
void reverseArray(int arr[], int start, int end) {
    for (int i = start, j = end; i < j; i++, j--) {
        swap(arr[i], arr[j]);
    }
}

// Function for left rotation by 'd' positions
void leftRotate(int arr[], int n, int d) {
    d = d % n;  // Handle cases where d > n

    // Step 1: Reverse first 'd' elements
    reverseArray(arr, 0, d - 1);
    // Step 2: Reverse remaining 'n-d' elements
    reverseArray(arr, d, n - 1);
    // Step 3: Reverse entire array
    reverseArray(arr, 0, n - 1);
}

// Function for right rotation by 'd' positions
void rightRotate(int arr[], int n, int d) {
    d = d % n;  // Handle cases where d > n

    // Step 1: Reverse last 'd' elements
    reverseArray(arr, n - d, n - 1);
    // Step 2: Reverse first 'n-d' elements
    reverseArray(arr, 0, n - d - 1);
    // Step 3: Reverse entire array
    reverseArray(arr, 0, n - 1);
}

// Function to print the array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int main() {
    int arr[] = {1, 2, 3, 4, 5, 6, 7};
    int n = sizeof(arr) / sizeof(arr[0]);
    int d;

    cout << "Enter the number of positions to rotate: ";
    cin >> d;

    int leftArr[n], rightArr[n];
    
    // Copy original array to perform rotations
    for (int i = 0; i < n; i++) {
        leftArr[i] = arr[i];
        rightArr[i] = arr[i];
    }

    // Perform rotations
    leftRotate(leftArr, n, d);
    rightRotate(rightArr, n, d);

    cout << "Array after left rotation by " << d << " positions: ";
    printArray(leftArr, n);

    cout << "Array after right rotation by " << d << " positions: ";
    printArray(rightArr, n);

    return 0;
}
